### ISSUE TITLE
performance issues on page load

### ISSUE THREAD

**Issue by garrett** [martinpitt/performance-graphs#51]
I'm seeing this oddity in Firefox, where it takes almost a minute to load the performance page _without graphs_ — this is just the current information on my machine.

Screencast:
http://people.redhat.com/~glesage/tmp/cockpit/Screencast%20from%2022-09-20%2016:53:45.webm

1. slow loading :00 - :27
2. page unresponsive banner between :27 and :45
3. blast of updates after :45, until around :50

**Comment by garrett** [comment]
This is in addition to the ouia issues, which #50 addresses.

(I've upgraded it here locally; this is why you don't see the ouia ID updates on the video.)

**Comment by martinpitt** [comment]
Some notekeeping from IRC:

 * I can't reproduce this even in the slowest possible situation that I can imagine: debug build, with pcp history, in a VM, with inspector open. After about 8s everything is rendered and constantly updating. Without the inspector, it's about 3s.

 * This seems to go away when Garrett disables cockpit-ostree.

 * Disabling all the timers does not make a difference (that would be really surprising, too, as the damage is already done before).

Does it make a differerence if you slow down the metrics update? It's in src/app.jsx line 227, here:

     this.metrics_channel = cockpit.channel({ payload: "metrics1", source: "internal", interval: 3000, metrics: CURRENT_METRICS });

You can try to bump 3000 to e.g. 30000 (once every 30s) and see if that makes any difference?

Also,  if you do `window.debugging = 'metrics` in the developer console (Ctrl+Shift+K), you will get lots of messages about metrics updates, etc. do you see something weird going on around the time when the CPU burst/hang happens, and when metrics are supposed to update? (every 3s by default, longer if you bump it as above).

**Comment by garrett** [comment]
> Garrett uses version 81.

Specifically the official Mozilla Firefox 81 from Flatpak. But the same problem happens in Chrome, WebKit, Firefox Nightly, etc...

**Comment by garrett** [comment]
Firefox's performance tab in the devtools says it's stuck in a setTimeout() in ostree.js. Chrome's performance tab says... nothing, as it hangs on trying to process the information it collected.

**Comment by andreasn** [comment]
I am unable to replicate this issue. Performance graphs page loads pretty much instantly for me. There was only a slight delay in loading the in/out values for Network.
Screencast:
https://andreasn.portknox.net/s/LzBXeWGMWDzkbni

The environment I used is the Fedora 32 image created with `make vm` in the performance-graphs repo.
Tested with Firefox 81 (flatpak), so it's not the browser. Chrome 85 is about the same.

**Comment by garrett** [comment]
Here's the performance view in devtools, with the frozen time highlighted, showing that it's definitely inside of a setTimeout... but coming from cockpit.js:

![performance-ostree-stack](https://user-images.githubusercontent.com/10246/94141396-f2b39580-fe6c-11ea-9c21-b7e99750ebed.png)

I switched views to the call tree and see it's from ostree.js:

![performance-ostree](https://user-images.githubusercontent.com/10246/94141693-6eaddd80-fe6d-11ea-961f-38708f30d960.png)

Expanding "JIT" (the largest timesink at the top), I see this:

![jit-ostree](https://user-images.githubusercontent.com/10246/94141957-cf3d1a80-fe6d-11ea-87dd-8baa84498c49.png)

...which also all comes from ostree.js.

**Comment by garrett** [comment]
> I am unable to replicate this issue.

> The environment I used is the Fedora 32 image

Right. This turns out _mainly_ to a problem with ostree, as stated in IRC and in the comments above. If you're not using a system with ostree, then you're not going to hit the big problem... and you're probably not going to notice the smaller one (which always happens).

The larger problem does show that updating the performance page is not done correctly, especially more noticable when there's a delay. (As it blasts the information instead of doing timely updates.)

> Tested with Firefox 81 (flatpak), so it's not the browser. Chrome 85 is about the same.

Yep, I mentioned in IRC and in the comments above, that I also see the problem with Firefox Nightly, GNOME Web, and Chrome too.
